File: test_api.js
const axios = require('axios');
const fs = require('fs');

// The API endpoint from SimpleQueryTool.vue
const API_URL = "https://unpaywall-simple-query-tool-c76bcdcacd9a.herokuapp.com/v2/dois";

// Sample valid DOIs from the test_dois.txt file
const validDois = [
  "10.1037/018882",
  "10.1086/466942",
  "10.1186/gb-2004-5-10-r80",
  "10.1080/00273171.2011.568786",
  "10.1016/j.cell.2007.11.019"
];

// Some malformed DOIs to test error handling
const malformedDois = [
  "10.1234/invalid",
  "not-a-doi",
  "10.1037/018882.invalid",
  "10.1",
  ""
];

// Combined test set
const allDois = [...validDois, ...malformedDois];

// Function to get response in CSV format
async function testCSVFormat() {
  console.log("Testing CSV format...");
  try {
    const response = await axios.post(API_URL, { dois: allDois }, {
      responseType: 'arraybuffer',
      timeout: 60000
    });
    
    // Convert ArrayBuffer to string
    const decoder = new TextDecoder('utf-8');
    const data = decoder.decode(response.data);
    
    fs.writeFileSync('api_response_csv.txt', data);
    console.log("CSV response saved to api_response_csv.txt");
    return data;
  } catch (error) {
    console.error("Error with CSV request:", error.message);
    if (error.response) {
      console.error("Response status:", error.response.status);
      if (error.response.data) {
        try {
          const decoder = new TextDecoder('utf-8');
          const errorText = decoder.decode(error.response.data);
          console.error("Error response:", errorText);
          fs.writeFileSync('api_error_csv.txt', errorText);
          console.log("CSV error response saved to api_error_csv.txt");
        } catch (e) {
          console.error("Could not decode error response");
        }
      }
    }
    return null;
  }
}

// Function to get response in JSONL format
async function testJSONLFormat() {
  console.log("Testing JSONL format...");
  try {
    const response = await axios.post(API_URL, { dois: allDois }, {
      responseType: 'arraybuffer',
      params: { format: 'jsonl' },
      timeout: 60000
    });
    
    // Convert ArrayBuffer to string
    const decoder = new TextDecoder('utf-8');
    const data = decoder.decode(response.data);
    
    fs.writeFileSync('api_response_jsonl.txt', data);
    console.log("JSONL response saved to api_response_jsonl.txt");
    return data;
  } catch (error) {
    console.error("Error with JSONL request:", error.message);
    if (error.response) {
      console.error("Response status:", error.response.status);
      if (error.response.data) {
        try {
          const decoder = new TextDecoder('utf-8');
          const errorText = decoder.decode(error.response.data);
          console.error("Error response:", errorText);
          fs.writeFileSync('api_error_jsonl.txt', errorText);
          console.log("JSONL error response saved to api_error_jsonl.txt");
        } catch (e) {
          console.error("Could not decode error response");
        }
      }
    }
    return null;
  }
}

// Test individual DOI to see detailed response
async function testSingleDoi(doi) {
  console.log(`Testing single DOI: ${doi}...`);
  try {
    const response = await axios.post(API_URL, { dois: [doi] }, {
      responseType: 'arraybuffer',
      params: { format: 'jsonl' },
      timeout: 60000
    });
    
    // Convert ArrayBuffer to string
    const decoder = new TextDecoder('utf-8');
    const data = decoder.decode(response.data);
    
    fs.writeFileSync(`api_response_${doi.replace('/', '_')}.txt`, data);
    console.log(`Response for DOI ${doi} saved to api_response_${doi.replace('/', '_')}.txt`);
    return data;
  } catch (error) {
    console.error(`Error with DOI ${doi}:`, error.message);
    if (error.response) {
      console.error("Response status:", error.response.status);
      if (error.response.data) {
        try {
          const decoder = new TextDecoder('utf-8');
          const errorText = decoder.decode(error.response.data);
          console.error("Error response:", errorText);
          fs.writeFileSync(`api_error_${doi.replace('/', '_')}.txt`, errorText);
          console.log(`Error for DOI ${doi} saved to api_error_${doi.replace('/', '_')}.txt`);
        } catch (e) {
          console.error("Could not decode error response");
        }
      }
    }
    return null;
  }
}

// Run tests
async function runTests() {
  // Test batch requests
  await testCSVFormat();
  await testJSONLFormat();
  
  // Test a few individual DOIs
  await testSingleDoi(validDois[0]);
  await testSingleDoi(malformedDois[0]);
  
  console.log("All tests completed!");
}

runTests();

File: test_api.py
import requests
import json
import os

# The API endpoint from SimpleQueryTool.vue
API_URL = "https://unpaywall-simple-query-tool-c76bcdcacd9a.herokuapp.com/v2/dois"

# Create a directory to store API responses if it doesn't exist
os.makedirs("api_responses", exist_ok=True)

# Sample valid DOIs from the test_dois.txt file
valid_dois = [
    "10.1037/018882",
    "10.1086/466942",
    "10.1186/gb-2004-5-10-r80",
    "10.1080/00273171.2011.568786",
    "10.1016/j.cell.2007.11.019"
]

# Some malformed DOIs to test error handling
malformed_dois = [
    "10.1234/invalid",
    "not-a-doi",
    "10.1037/018882.invalid",
    "10.1",
    ""
]

# Combined test set
all_dois = valid_dois + malformed_dois

def test_csv_format():
    """Test the API with CSV format response"""
    print("Testing CSV format...")
    try:
        response = requests.post(API_URL, json={"dois": all_dois}, timeout=60)
        
        # Save the response
        with open("api_responses/response_csv.txt", "w") as f:
            f.write(response.text)
        
        print(f"CSV response saved to api_responses/response_csv.txt")
        return response.text
    except Exception as e:
        print(f"Error with CSV request: {str(e)}")
        if hasattr(e, 'response') and e.response:
            print(f"Response status: {e.response.status_code}")
            with open("api_responses/error_csv.txt", "w") as f:
                f.write(e.response.text)
            print(f"CSV error response saved to api_responses/error_csv.txt")
        return None

def test_jsonl_format():
    """Test the API with JSONL format response"""
    print("Testing JSONL format...")
    try:
        response = requests.post(API_URL, json={"dois": all_dois}, params={"format": "jsonl"}, timeout=60)
        
        # Save the response
        with open("api_responses/response_jsonl.txt", "w") as f:
            f.write(response.text)
        
        print(f"JSONL response saved to api_responses/response_jsonl.txt")
        return response.text
    except Exception as e:
        print(f"Error with JSONL request: {str(e)}")
        if hasattr(e, 'response') and e.response:
            print(f"Response status: {e.response.status_code}")
            with open("api_responses/error_jsonl.txt", "w") as f:
                f.write(e.response.text)
            print(f"JSONL error response saved to api_responses/error_jsonl.txt")
        return None

def test_single_doi(doi):
    """Test the API with a single DOI"""
    print(f"Testing single DOI: {doi}...")
    try:
        response = requests.post(API_URL, json={"dois": [doi]}, params={"format": "jsonl"}, timeout=60)
        
        # Create a valid filename
        filename = doi.replace("/", "_").replace(":", "_")
        
        # Save the response
        with open(f"api_responses/response_{filename}.txt", "w") as f:
            f.write(response.text)
        
        print(f"Response for DOI {doi} saved to api_responses/response_{filename}.txt")
        return response.text
    except Exception as e:
        print(f"Error with DOI {doi}: {str(e)}")
        if hasattr(e, 'response') and e.response:
            print(f"Response status: {e.response.status_code}")
            with open(f"api_responses/error_{filename}.txt", "w") as f:
                f.write(e.response.text)
            print(f"Error for DOI {doi} saved to api_responses/error_{filename}.txt")
        return None

def create_combined_response_file():
    """Create a single file with all responses for easier reference"""
    print("Creating combined response file...")
    
    content = "# API Test Results\n\n"
    
    # Add CSV format results
    content += "## CSV Format (All DOIs)\n\n"
    try:
        with open("api_responses/response_csv.txt", "r") as f:
            csv_content = f.read()
        content += "```csv\n" + csv_content + "\n```\n\n"
    except:
        content += "CSV response not available\n\n"
    
    # Add JSONL format results
    content += "## JSONL Format (All DOIs)\n\n"
    try:
        with open("api_responses/response_jsonl.txt", "r") as f:
            jsonl_content = f.read()
            
        # Try to pretty format each JSON line
        pretty_jsonl = ""
        for line in jsonl_content.strip().split("\n"):
            try:
                obj = json.loads(line)
                pretty_jsonl += json.dumps(obj, indent=2) + "\n\n"
            except:
                pretty_jsonl += line + "\n"
                
        content += "```json\n" + pretty_jsonl + "\n```\n\n"
    except:
        content += "JSONL response not available\n\n"
    
    # Add individual DOI results
    content += "## Individual DOI Tests\n\n"
    
    # Valid DOIs
    content += "### Valid DOIs\n\n"
    for doi in valid_dois:
        filename = doi.replace("/", "_").replace(":", "_")
        content += f"#### DOI: {doi}\n\n"
        try:
            with open(f"api_responses/response_{filename}.txt", "r") as f:
                response_content = f.read()
                
            # Try to pretty format JSON
            try:
                obj = json.loads(response_content)
                pretty_json = json.dumps(obj, indent=2)
                content += "```json\n" + pretty_json + "\n```\n\n"
            except:
                content += "```\n" + response_content + "\n```\n\n"
        except:
            content += f"Response not available for {doi}\n\n"
    
    # Malformed DOIs
    content += "### Malformed DOIs\n\n"
    for doi in malformed_dois:
        if not doi:  # Skip empty string
            doi_display = "(empty string)"
            filename = "empty"
        else:
            doi_display = doi
            filename = doi.replace("/", "_").replace(":", "_")
            
        content += f"#### DOI: {doi_display}\n\n"
        try:
            with open(f"api_responses/response_{filename}.txt", "r") as f:
                response_content = f.read()
                
            # Try to pretty format JSON
            try:
                obj = json.loads(response_content)
                pretty_json = json.dumps(obj, indent=2)
                content += "```json\n" + pretty_json + "\n```\n\n"
            except:
                content += "```\n" + response_content + "\n```\n\n"
        except:
            content += f"Response not available for {doi_display}\n\n"
    
    # Save the combined file
    with open("api_responses.md", "w") as f:
        f.write(content)
        
    print("Combined response saved to api_responses.md")

def run_tests():
    """Run all tests"""
    # Test batch requests
    test_csv_format()
    test_jsonl_format()
    
    # Test individual valid DOIs
    for doi in valid_dois:
        test_single_doi(doi)
    
    # Test individual malformed DOIs
    for doi in malformed_dois:
        if doi:  # Skip empty string
            test_single_doi(doi)
    
    # Create combined response file
    create_combined_response_file()
    
    print("All tests completed!")

if __name__ == "__main__":
    run_tests()
